File: api/routers/scoring.py
"""Score calculation endpoints."""
from fastapi import APIRouter, Query
from pydantic import BaseModel
from src.scoring import get_ring_score, get_flint_score
import math

router = APIRouter()


class ScoreResult(BaseModel):
    """Result of a score calculation."""
    score: int
    is_x: bool = False


@router.get("/ring", response_model=ScoreResult)
def calculate_ring_score(
    x: float = Query(..., description="X coordinate on target face (cm)"),
    y: float = Query(..., description="Y coordinate on target face (cm)"),
    face_cm: int = Query(..., description="Face size in cm (e.g., 40, 60, 80, 122)"),
    face_type: str = Query(default="WA", description="Face type: 'WA' or 'Flint'"),
    x_is_11: bool = Query(default=False, description="Whether X ring counts as 11 (compound scoring)")
):
    """
    Calculate score from shot coordinates.
    
    Returns the score and whether it's an X (inner 10).
    """
    # Calculate radius from center
    radius_cm = math.sqrt(x**2 + y**2)
    
    if face_type.upper() == "FLINT":
        score = get_flint_score(radius_cm, face_cm)
        is_x = False  # Flint doesn't have X ring scoring
    else:
        score = get_ring_score(radius_cm, face_cm, x_is_11)
        # Check if it's an X (inner 10)
        ring_width = face_cm / 20.0
        x_radius = ring_width / 2
        is_x = (score >= 10 and radius_cm <= x_radius)
    
    return ScoreResult(score=score, is_x=is_x)

File: frontend/src/domain/scoring.ts
/**
 * Scoring functions for WA and IFAA Flint target faces.
 * Ported from src/scoring.py — ring-score calculation only.
 * create_target_face() is NOT ported (Plotly rendering stays in the frontend components).
 */

/**
 * Calculate score based on distance from center for WA target faces (1-10 rings).
 * Ring width = face diameter / 20.
 */
export function getRingScore(radiusCm: number, faceSizeCm: number, xIs11 = false): number {
  const ringWidth = faceSizeCm / 20.0;

  if (ringWidth <= 0) return 0;
  if (radiusCm < 0) return 0;

  const ringIndex = Math.ceil(radiusCm / ringWidth);

  if (ringIndex <= 1) {
    if (xIs11 && radiusCm <= ringWidth / 2) return 11;
    return 10;
  } else if (ringIndex <= 10) {
    return 11 - ringIndex;
  }
  return 0; // Miss
}

/**
 * Calculate score for IFAA Flint/Field rounds (5, 4, 3).
 * 5-ring diameter = face_size * 0.2
 * 4-ring diameter = face_size * 0.4
 * 3-ring diameter = face_size * 0.6
 */
export function getFlintScore(radiusCm: number, faceSizeCm: number): number {
  const r5 = (faceSizeCm * 0.2) / 2;
  const r4 = (faceSizeCm * 0.4) / 2;
  const r3 = (faceSizeCm * 0.6) / 2;

  if (radiusCm <= r5) return 5;
  if (radiusCm <= r4) return 4;
  if (radiusCm <= r3) return 3;
  return 0;
}
